var/spack/repos/builtin/packages/glib/package.py

```diff
@@ -180,6 +190,86 @@ def patch(self):
     def libs(self):
         return find_libraries(["libglib*"], root=self.prefix, recursive=True)
 
+    @property
+    def dtrace_copy_path(self):
+        return join_path(self.stage.source_path, "dtrace-copy")
+
+    @run_before("install")
+    def fix_python_path(self):
+        if not self.spec.satisfies("@2.53.4:"):
+            return
+
+        files = ["gobject/glib-genmarshal.in", "gobject/glib-mkenums.in"]
+
+        filter_file(
+            "^#!/usr/bin/env @PYTHON@",
+            "#!/usr/bin/env {0}".format(os.path.basename(self.spec["python"].command.path)),
+            *files,
+        )
+
+    @run_before("install")
+    def fix_dtrace_usr_bin_path(self):
+        if "tracing=dtrace" not in self.spec:
+            return
+
+        # dtrace may cause glib build to fail because it uses
+        # '/usr/bin/python' in the shebang. To work around that
+        # we copy the original script into a temporary folder, and
+        # change the shebang to '/usr/bin/env python'
+        dtrace = which("dtrace").path
+        dtrace_copy = join_path(self.dtrace_copy_path, "dtrace")
+
+        with working_dir(self.dtrace_copy_path, create=True):
+            copy(dtrace, dtrace_copy)
+            filter_file(
+                "^#!/usr/bin/python",
+                "#!/usr/bin/env {0}".format(os.path.basename(self.spec["python"].command.path)),
+                dtrace_copy,
+            )
+
+        # To have our own copy of dtrace in PATH, we need to
+        # prepend to PATH the temporary folder where it resides
+        env["PATH"] = ":".join([self.dtrace_copy_path] + env["PATH"].split(":"))
+
+    @run_after("install")
+    def filter_sbang(self):
+        # Revert sbang, so Spack's sbang hook can fix it up (we have to do
+        # this after install because otherwise the install target will try
+        # to rebuild files as filter_file updates the timestamps)
+        if self.spec.satisfies("@2.53.4:"):
+            pattern = "^#!/usr/bin/env {0}".format(
+                os.path.basename(self.spec["python"].command.path)
+            )
+            repl = "#!{0}".format(self.spec["python"].command.path)
+            files = ["glib-genmarshal", "glib-mkenums"]
+        else:
+            pattern = "^#! /usr/bin/perl"
+            repl = "#!{0}".format(self.spec["perl"].command.path)
+            files = ["glib-mkenums"]
+
+        files = [join_path(self.prefix.bin, file) for file in files]
+        filter_file(pattern, repl, *files, backup=False)
```

Comment:
Functions decorated with `run_after` and `run_before` cannot stay in the `*Package` when the recipe supports multiple build-systems. I think these functions are currently never run.

Reply:
Thanks for catching this! Is there a nicer way than simply copying all these methods into both `Builder` classes (which seems to work)?

Reply:
@michaelkuhn You can do it like it is done here https://github.com/spack/spack/pull/36825/files in the `netcdf-c` package. Insert the base builder last in the MRO and everything should be fine.

Reply:
Thanks for the hint! That was easier than I thought. :smile: I've pushed a new version that should now work correctly.